### PR TITLE
Adding `integration:auth` target

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR adds a `mage integration:auth` target.  For starters, it just implements authentication against GCP and ESS (QA).

## Why is it important?

This target will be invoked either directly by the user or indirectly by the integration testing framework to authenticate the user to various IaaS CSPs (GCP, AWS, Azure) and ESS (QA).   The framework will need this authentication so it can provision the necessary resources (VMs, Elastic stack deployments) as needed for the integration/E2E tests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [ ] ~I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~
- [ ] ~I have added an integration test or an E2E test~

### Pre-requisites

* [Install](https://cloud.google.com/sdk/docs/install) the `gcloud` CLI.

### Without being authenticated to either `gcp` or ESS

1. Run `mage integration:auth`
2. First, you should be taken through the GCP authentication workflow, which includes opening your browser and signing in with your `@elastic.co` Google account.  When you are done in the browser, return to your terminal.
3. Next, you should be taken through the ESS authentication workflow, which includes being prompted on the command-line for your ESS API key.  If you don't have one already, [create one](https://www.elastic.co/guide/en/cloud/current/ec-api-authentication.html) in the [ESS **QA** User Console UI](https://console.qa.cld.elstc.co/).

### After being authenticated to both `gcp` and ESS

1. Run `mage integration:auth`
2. You should not be taken through any authentication workflows.  Instead, you should just see messages on STDOUT saying you are authenticated to GCP and ESS.

### Being authenticated to either `gcp` or ESS but not both
1. Run `mage integration:auth`
2. For the service you are not authenticated to, you should be taken through the authentication workflow as described in the first scenario above.
3. For the service you are authenticated to, you should not be taken through any authentication workflows.  Instead, you should just see a message on STDOUT saying you are authenticated to that service.

#### Tips 
* To un-authenticate with GCP, run `gcloud auth revoke`
* To un-authenticate with ESS, run one of these:
   * `rm ~/.config/ess/api_key.txt` — simulates the "first use" scenario
   * `echo -n $RANDOM$RANDOM > ~/.config/ess/api_key.txt` — simulates the ESS API key being bogus or revoked

### CODE REVIEW

File: magefile.go [hunk 1]
```diff
@@ -1284,6 +1290,26 @@ func (Integration) Local(ctx context.Context) error {
...
+	if err := authGCP(ctx); err != nil {
+		return fmt.Errorf("unable to authenticate to GCP: %w", err)
+	}
+	fmt.Println("✔️  GCP authentication successful")
+
+	// TODO: Authenticate user to AWS
+
+	// TODO: Authenticate user to Azure
+
+	// TODO: Authenticate user to ESS
```
Comment (blakerouse) [inline magefile.go:1304]: This should be removed, correct?

File: magefile.go [hunk 2]
```diff
@@ -1295,3 +1321,123 @@ func shouldBuildAgent() bool {
...
+	// Check if gcloud CLI is installed
+	const cliName = "gcloud"
+	cmd := exec.CommandContext(ctx, "which", cliName)
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("%s CLI is not installed: %w", cliName, err)
+	}
+
+	// Check if user is already authenticated
+	for authSuccess := false; !authSuccess; {
+		cmd = exec.CommandContext(ctx, cliName, "auth", "list", "--filter=status:ACTIVE", "--format=json")
```
Comment (blakerouse) [inline magefile.go:1336]: Should we check (if possible) to determine that they are authenticated with an elastic account that they are using the correct platform ingest project?

File: magefile.go [hunk 3]
```diff
@@ -1295,3 +1321,123 @@ func shouldBuildAgent() bool {
...
+		essAPIKey, err = stringPrompt("Please provide a ESS (QA) API key:")
+		if err != nil {
+			return fmt.Errorf("unable to read ESS API key from prompt: %w", err)
+		}
+	}
+
+	// Write API key to file for future use
+	if err := os.WriteFile(essAPIKeyFile, []byte(essAPIKey), 0600); err != nil {
+		return fmt.Errorf("unable to persiste ESS API key for future use: %w", err)
+	}
```
Comment (blakerouse) [inline magefile.go:1421]: I think we should validate that the provided key actually works at this point.

File: magefile.go (hunk 2 above)
Comment (ycombinator) [reply magefile.go:1336]: Yes, good idea.  I think we can at least check the domain of the email address they're authenticated with, which should be good enough to say if they're authenticated with an elastic account.  I'm not sure if there's a better / more structured way than that.

I think we can check the GCP project as well.  Do you recall which project we're supposed to be using?  If I sign in, I see `elastic-obs-integrations-dev` as one of the options for me.  I think that's it but maybe there's another one for Ingest?  What do you see when you sign in?

Comment (ycombinator) [reply magefile.go:1335]: Thinking about this some more, do we want to add these extra checks, considering this target is OSS?  Say an external (non-Elastic) contributor wanted to write an integration test and wanted to run their test locally, these restrictions would get in their way.  Also, as far as the test runner is concerned, does it need to really care about whether the authenticated account is an `@elastic.co` email address or if it's using the `elastic-obs-integrations-dev` GCP project?

File: magefile.go (hunk 3 above)
Comment (ycombinator) [reply magefile.go:1420]: I'm not sure I understand.  We're doing that as the very first thing in the loop.  I'll clarify the comment a bit more in the success conditional inside the loop.

Comment (ycombinator) [reply magefile.go:1441]: Done in eb345a605fcd5d0e01bb129efd0f9869ce9d5cf2.

File: magefile.go (hunk 2 above)
Comment (ycombinator) [reply magefile.go:1338]: I implemented the extra checks in 029b8d59ca5476e35d875a673def0e62e282ee00 but I'm really not sure we should add these, given that this is an OSS project repo (see previous comment for more detail).  WDYT?

Comment (blakerouse) [reply magefile.go:1338]: I am torn. I think it will be helpful for us, and we will be the main users. I also don't know how many outside of Elastic will want to run integration tests on there own GCP account, feel like they will more likely rely on CI. But being an OSS project I could see that we might not want it. I am on the fence.

File: magefile.go [hunk 4]
```diff
@@ -1295,3 +1320,144 @@ func shouldBuildAgent() bool {
...
+
+	// Check the authenticated account's project
+	cmd = exec.CommandContext(ctx, cliName, "config", "get", "core/project")
+	output, err := cmd.Output()
+	if err != nil {
+		return fmt.Errorf("unable to get project: %w", err)
+	}
+
+	const expectedProject = "elastic-obs-integrations-dev"
```
Comment (blakerouse) [inline magefile.go:1378]: I believe this should be `elastic-platform-ingest`.

**Review by blakerouse — DISMISSED** [review]

File: magefile.go (hunk 4 above)
Comment (pierrehilbert) [reply magefile.go:1378]: I agree, we should use elastic-platform-ingest

Comment (ycombinator) [reply magefile.go:1378]: Thanks, will change.  @pierrehilbert How do I get access to this project, so I can test this PR?

File: magefile.go (hunk 2 above)
Comment (ycombinator) [reply magefile.go:1338]: How about we expose a boolean environment variable, something like `TEST_INTEG_AUTH_GCP_SKIP_ELASTIC_CHECK`.  Unless this variable is set to a `true`-y value (basically anything [`strconv.ParseBool`](https://pkg.go.dev/strconv#ParseBool) would evaluate to `true`), we'd perform the additional Elastic-specific checks. 

In other words, by default, the Elastic-specific checks will be performed but if a non-Elastic contributor wants to run this target locally, they have a way to skip past these Elastic-specific checks.

WDYT?

File: magefile.go (hunk 4 above)
Comment (jlind23) [reply magefile.go:1378]: @ycombinator you need to be part of the ingest-team google group which is already the case.

File: magefile.go (hunk 2 above)
Comment (ycombinator) [reply magefile.go:1338]: I implemented the skip env var idea (see previous comment) in 10eeeea9d.  Let me know what you think.

File: magefile.go [hunk 5]
```diff
@@ -1295,3 +1320,170 @@ func shouldBuildAgent() bool {
 	}
 	return ret
 }
+
+// Pre-requisite: user must have the gcloud CLI installed
+func authGCP(ctx context.Context) error {
+	// Check if gcloud CLI is installed
+	const cliName = "gcloud"
+	cmd := exec.CommandContext(ctx, "which", cliName)
```
Comment (pchila) [inline magefile.go:1328]: How would this work on windows? Are we not going to support devs that want to run integration tests on Windows? They should perform auth manually ? Should we use golang libraries to authenticate against gcloud/aws/azure APIs ?

File: magefile.go [hunk 6]
```diff
@@ -1295,3 +1320,170 @@ func shouldBuildAgent() bool {
...
+
+		// Try to authenticate user
+		cmd = exec.CommandContext(ctx, cliName, "auth", "login")
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("unable to authenticate user: %w", cliName, err)
+		}
+	}
+
+	// Skip additional Elastic-specific checks if so requested
+	const skipEnvVar = "TEST_INTEG_AUTH_GCP_SKIP_ELASTIC_CHECK"
```
Comment (pchila) [inline magefile.go:1364]: Maybe we can use an env var or an argument to the mage target to specify a a project to switch to instead of disabling the check on the email and project switching...
This should also make it easy to create different launch configs for IDEs etc.

File: magefile.go [hunk 7]
```diff
@@ -1295,3 +1320,170 @@ func shouldBuildAgent() bool {
...
+
+	return nil
+}
+
+func authESS(ctx context.Context) error {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return fmt.Errorf("unable to determine user's home directory: %w", err)
+	}
+	essAPIKeyFile := filepath.Join(homeDir, ".config", "ess", "api_key.txt")
```
Comment (pchila) [inline magefile.go:1418]: Is this location portable? (I don't know if it's the right place to store it in Windows)

File: magefile.go [hunk 8]
```diff
@@ -1295,3 +1320,170 @@ func shouldBuildAgent() bool {
...
+	}
+
+	// Read API key from file
+	data, err := os.ReadFile(essAPIKeyFile)
+	if err != nil {
+		return fmt.Errorf("unable to read ESS API key: %w", err)
+	}
+
+	data = bytes.TrimSpace(data)
+	essAPIKey := string(data)
```
Comment (pchila) [inline magefile.go:1440]: Since we know it's all a string what about:
```suggestion
	essAPIKey := strings.TrimSpace(string(data))
```

File: pkg/testing/ess/users.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+type GetUserResponse struct {
+	User struct {
+		UserID int `json:"user_id"`
+	} `json:"user"`
+}
+
+// GetUser returns information about the authenticated user
+func (c *Client) GetUser(ctx context.Context, req GetUserRequest) (*GetUserResponse, error) {
```
Comment (pchila) [inline pkg/testing/ess/users.go:24]: Nitpicking: as a preference I like to see the methods close to the types that they refer to... Why is this here and not in `client.go` ?

**Review by pchila — COMMENTED** [review]
Aside from a couple of style preferences, I would like to see if we can connect to gcp also on windows, maybe it's easier if we use some gcp dependencies instead of relying on `gcloud` CLI

File: magefile.go (hunk 5 above)
Comment (ycombinator) [reply magefile.go:1328]: I think `gcloud` (the CLI) is [installable on Windows](https://cloud.google.com/sdk/docs/install#windows) too but, yes, we'd need to replace this `which` usage here — good catch.  

We could also consider using golang APIs but I'm not sure what downstream implications that has for the test runner.  I think we are currently working under the assumption that it will depend on the config/profile files created by these CLIs for auth tokens, selected projects, etc.

Let me start by trying to implementing the equivalent of `which` here for Windows.  If that doesn't do it, then we can consider the more involved option of replacing the CLIs with golang libraries.

File: magefile.go (hunk 6 above)
Comment (ycombinator) [reply magefile.go:1364]: I like this idea - will do!

File: magefile.go (hunk 7 above)
Comment (ycombinator) [reply magefile.go:1418]: It should be because `homeDir` itself should be portable (see [`os.UserHomeDir()`](https://pkg.go.dev/os#UserHomeDir)).

File: pkg/testing/ess/users.go (hunk 9 above)
Comment (ycombinator) [reply pkg/testing/ess/users.go:24]: Mostly because the ESS API is quite large.  I don't think we'd every fully implement it but, if we did, `client.go` might get quite large.

Nevertheless, your point is well taken.  We should start by keeping everything in `client.go`.  If it starts to get really large, we can consider breaking it up.

File: magefile.go (hunk 2 above)
Comment (blakerouse) [reply magefile.go:1338]: Yeah. I like the env variable.

File: magefile.go (hunk 5 above)
Comment (blakerouse) [reply magefile.go:1328]: This should work on Windows, absolutely. I think we can push that to a follow up PR.

File: pkg/testing/ess/users.go (hunk 9 above)
Comment (blakerouse) [reply pkg/testing/ess/users.go:24]: I like the idea of keeping each area of the API in its own file, so please leave it like it is.

**Review by blakerouse — DISMISSED** [review]
I say we merge this and iterate.

File: magefile.go (hunk 5 above)
Comment (ycombinator) [reply magefile.go:1328]: I think the issue is with the `which` command, which doesn't exist on Windows.  But there is a `where` command in Windows that does the same thing, apparently, so I'll just use that here for Windows.

Comment (ycombinator) [reply magefile.go]: Implemented the `which`/`where` change.  

As for the CLI vs. golang APIs, ++ to deferring this decision to a follow up PR.

File: magefile.go (hunk 6 above)
Comment (ycombinator) [reply magefile.go]: Done in caeecb1ec.

**Comment by ycombinator** [comment]
@blakerouse @pchila Thanks for your review.  I've addressed all your feedback now.  Please re-review when you get a chance.

**Review by blakerouse — APPROVED** [review]
Nice! Looks good.

**Review by pchila — APPROVED** [review]
LGTM
